webapp/src/snippetBuilder.tsx

```diff
@@ -0,0 +1,339 @@
+        const defaults: any = {};
+
+        for (const question of config.questions) {
+            const { inputs } = question;
+            for (const input of inputs) {
+                const { defaultAnswer, answerToken } = input;
+                defaults[answerToken] = defaultAnswer;
+            }
+        }
+
+        this.setState({ defaults });
+    }
+
+    componentDidMount() {
+        // Sets default values
+        this.buildDefaults();
+    }
+
+    replaceTokens(output: string) {
+        const { answers, defaults } = this.state;
+        let tokenizedOutput = output;
+        const tokens = Object.keys(defaults);
+
+        // Replaces output tokens with answer if available or default value
+        for (let token of tokens) {
+            if (answers[token]) {
+                tokenizedOutput = tokenizedOutput.split(`$${token}`).join(answers[token]);
+            }
+            else {
+                tokenizedOutput = tokenizedOutput.split(`$${token}`).join(defaults[token]);
+            }
+        }
+
+        return tokenizedOutput;
+    }
+
+    generateOutputMarkdown(output: string) {
+        const { config } = this.state;
+        // Attaches starting and ending line based on output type
+        let md = `\`\`\`${config.outputType}\n`;
+        md += this.replaceTokens(output);
+        md += `\n\`\`\``;
+
+        return md
+    }
+
+    hide() {
+        this.setState({
+            visible: false
+        });
+    }
+
+    show(projectView: pxt.editor.IProjectView, mainWorkspace: Blockly.Workspace) {
+        pxt.tickEvent('snippetBuilder.show', null, { interactiveConsent: false });
+        this.setState({
+            visible: true,
+            mainWorkspace,
+            projectView,
+        });
+    }
+
+    cancel() {
+        pxt.tickEvent("snippetBuilder.cancel", undefined, { interactiveConsent: true });
+        this.hide();
+    }
+
+    injectBlocksToWorkspace() {
+        const { mainWorkspace, output } = this.state;
+
+        compiler.getBlocksAsync()
+            .then(blocksInfo => compiler.decompileBlocksSnippetAsync(this.replaceTokens(output), blocksInfo))
+            .then(resp => {
+                Blockly.Xml.appendDomToWorkspace(Blockly.Xml.textToDom(resp), mainWorkspace);
+            });
+    }
+
+    confirm() {
+        this.injectBlocksToWorkspace();
+        this.hide();
+    }
+
+    changePage(increment: 1 | -1) {
+        const { currentQuestion } = this.state;
+        this.setState({ currentQuestion: currentQuestion + increment });
+    }
+
+    nextPage() {
+        const { config } = this.state;
+        const { currentQuestion, output } = this.state;
+        const nextQuestion = config.questions[currentQuestion + 1];
+        // If output exists
+        if (nextQuestion.output) {
+            // If output is not already appended
+            if (output.indexOf(nextQuestion.output) === -1) {
+                this.setState({ output: `${output}\n${nextQuestion.output}`});
```

Comment:
I suspect you might actually want to keep each question output separate and then only concatenate them at the very end. You could probably do this by making "output" a string[] instead of a string.
How do you handle revisiting old questions if it's all been combined into one output string?

Reply:
If a question is being revised I have to make sure that the tokens are replaced everywhere the answer correlates to a token. Since they're unique in nature running it over the entire output string is necessary. This could definitely be turned into an array and mapped over if you think it's better. It will have to be combined before being passed to MarkedContent regardless. Otherwise the first item in output[] would be the initial output, the second is the initial output and the new output, and so on.